File: 559.MaximumDepthofN-aryTree.js
/**
 * // Definition for a Node.
 * function Node(val,children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node} root
 * @return {number}
 */
var maxDepth = function(root) {
    if (root) {
        const childs = root.children.map(c => maxDepth(c));
        return 1 + (childs.length ? Math.max(...childs) : 0);
    }
    return 0;
};

var maxDepth = function(root) {
    const stack = [root];
    let depth = 0;
    let count = 1;
    let cursor = 0;
    while(stack.length) {
        const node = stack.shift();
        if (node) {
            node.children.forEach(c => stack.push(c));
            cursor++;
            if (cursor === count) {
                cursor = 0;
                count = stack.length;
                depth++;
            }
        }
    }
    return depth;
 };

File: Golang/Tree/559.Maximum-depth-of-n-ary-tree.go
package main

type Node struct {
	Val      int
	Children []*Node
}

func maxDepth(root *Node) int {
	var recursive func(node *Node) int
	recursive = func(node *Node) int {
		if node != nil {
			maxValue := 0
			for _, n := range node.Children {
				if currentDepth := recursive(n); currentDepth > maxValue {
					maxValue = currentDepth
				}
			}
			return maxValue + 1
		}
		return 0
	}
	return recursive(root)
}

func maxDepthBFS(root *Node) int {
	ret := 0
	if root != nil {
		stack := []*Node{root}
		for len(stack) > 0 {
			children := []*Node{}
			for i := 0; i < len(stack); i++ {
				node := stack[i]
				children = append(children, node.Children...)
			}
			ret++
			stack = children
		}
	}
	return ret
}
